### DISCUSSION TITLE
Quarkus - Oracle 19 connection issue

### DISCUSSION THREAD

**Discussion by chileme88** [quarkusio/quarkus#33989] [Q&A]
Hi all, can you help me with this strange issue? 

In my project I need to use Quarkus with Oracle 19c 

Quarkus configuration:
    Quarkus 2.16.7
    Driver jdbc:quarkus-jdbc-oracle (ojdbc11 21.5.0.0)
    Framework ORM: Hibernate (5.6.15) with Panache

jdbc configuration:
    quarkus.datasource.db-kind=oracle
    quarkus.datasource.jdbc.driver=oracle.jdbc.driver.OracleDriver
    quarkus.hibernate-orm.dialect=org.hibernate.dialect.Oracle12cDialect

Connection configuration:
    quarkus.datasource.jdbc.url=jdbc:oracle:thin:@//aaa:1521/bbb
    quarkus.datasource.username
    quarkus.datasource.password

Database configuration:
	Version: 19.17
	Pluggable database
	Single node
	
SQLNET.ORA SERVER:
	SQLNET.ENCRYPTION_SERVER=required
	NAMES.DIRECTORY_PATH=(TNSNAMES, EZCONNECT)
	SQLNET.CRYPTO_SEED='sasfafafsafasfsadsadasdadsada'
	SQLNET.ENCRYPTION_TYPES_SERVER=(AES256)
	TRACE_TIMESTAMP_CLIENT=ON
	ADR_BASE=/u01/oracle/app/db
	SQLNET.CRYPTO_CHECKSUM_SERVER=required
	SQLNET.CRYPTO_CHECKSUM_TYPES_SERVER=(SHA512, SHA1)

ORACLE SERVICE:
	Service name: PDBMWDEVSRV
	Cardinality: SINGLETON
	Service role: PRIMARY
	Management policy: AUTOMATIC
	DTP transaction: false
	AQ HA notifications: false
	Global: false
	Commit Outcome: true
	Failover type: AUTO
	Failover method:
	Failover retries: 1
	Failover delay: 3
	Failover restore: AUTO
	Connection Load Balancing Goal: LONG
	Runtime Load Balancing Goal: THROUGHPUT
	TAF policy specification: NONE
	Edition:
	Pluggable database name: PDBMWDEV
	Hub service:
	Maximum lag time: ANY
	SQL Translation Profile:
	Retention: 86400 seconds
	Replay Initiation Time: 600 seconds
	Drain timeout:
	Stop option: transactional
	Session State Consistency: AUTO
	GSM Flags: 0
	Service is enabled
	Service uses Java: false
	

When I start the pod, I can see an estabilished connection to the server:
	12-JUN-2023 15:24:21 * (CONNECT_DATA=(CID=(PROGRAM=JDBC Thin Client)(HOST=__jdbc__)(USER=1001560000))(SERVICE_NAME=PDBMWDEVSRV)(CONNECTION_ID=z9MY0UBhTBqddgtLw9xRxw==)) * (ADDRESS=(PROTOCOL=tcp)(HOST=10.10.32.8)(PORT=33756)) * establish * PDBMWDEVSRV * 0

but in the Quarkus logs I have:
	2023-06-12 15:24:48,285 WARN  [io.agr.pool] (agroal-11) Datasource '<default>': IO Error: Socket read interrupted, Authentication lapse 0 ms.

After this connection no session are created in the database and the next query fails.

This goes on for a number of retry on the same or different query calls, up until it suddely starts working
After the first succesfull query is executed, the app works fine for some time (even some days) until it runs again in the connection problem and I have to restart the pod

Do you can help me please? I don't understand what is happening.

**Comment by ccarv** [comment]
Did you ever manage to get past this error? We recently migrated an app from Quarkus 2.x to 3.x, and are now experiencing the same error.

We went as far as creating a simple test application using the same JDBC parameters and still receive the same error.

**Comment by solerman** [comment]
Came to these post by pure chance, I don't use (yet) Quarkus. Looking up  the error, there is this in metalink:

> CAUSE
> This is the issue investigated in unpublished / internal Bug 26414928 - SOCKET READ INTERRUPTED AND IO ERROR: OPERATION INTERRUPTED WITH OJDBC8 JARS

> This was closed as "Not a Bug".
 
> The 12.2.0.1 and above JDBC driver uses Java NIO calls in blocking mode, which can be impacted by any interrupt() calls being made by the application.

> This differs from previous versions of the JDBC driver, which used stream-based I/O API calls that were not affected by calls to interrupt().

> Note that this is a deliberate / intentional change beginning in the 12.2.0.1 JDBC driver, rather than a bug.
 

> SOLUTION
> Inspect the application code for any interrupt() method calls.

> Alternatively, set the following connection property:  oracle.jdbc.javaNetNio=false.

> Example:

> java -Doracle.jdbc.javaNetNio=false ...

>  This will direct the JDBC driver to revert to the  pre-12.2 behavior of using stream-based I/O API calls no affected by interrupt() method calls.

**Comment by LarsKoelpin** [comment]
`-Doracle.jdbc.javaNetNio=false` does not seem to be working in Quarkus 3 (at least when using keycloak 22.04).

``` 
Could not obtain connection to query metadata: java.sql.SQLRecoverableException: ORA-17002: I/O error: java.lang.NullPointerException: Cannot invoke "java.nio.channels.SocketChannel.write(java.nio.ByteBuffer)" because "this.session.socketChannel" is null, connect lapse 3 ms., Authentication lapse 0 ms.
files-keycloak-1               |        at oracle.jdbc.driver.T4CConnection.handleLogonIOException(T4CConnection.java:1142)
files-keycloak-1               |        at oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:882)
files-keycloak-1               |        at oracle.jdbc.driver.PhysicalConnection.connect(PhysicalConnection.java:1132)
files-keycloak-1               |        at oracle.jdbc.driver.T4CDriverExtension.getConnection(T4CDriverExtension.java:104)
files-keycloak-1               |        at oracle.jdbc.driver.OracleDriver.connect(OracleDriver.java:837)
files-keycloak-1               |        at oracle.jdbc.driver.OracleDriver.connect(OracleDriver.java:664)
files-keycloak-1               |        at io.agroal.pool.ConnectionFactory.createConnection(ConnectionFactory.java:226)
files-keycloak-1               |        at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:536)
files-keycloak-1               |        at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:517)
```

I am having a similar problem. Are there any further solutions?

**Comment by ccarv** [comment]
@LarsKoelpin we spent several weeks trying different solutions and ended up reverting to our 2.x codebase. Unfortunately, until the Quarkus team recognizes this as an issue, 2.x is the last version we'll be able to use.

Fingers crossed someone from Red Hat acknowledges this post!

**Comment by aniash007** [comment]
Upgrade the OJDBC8 to 23.3.0.23.09

**Comment by fugerit79** [comment]
Godd morning everyone .

We are running a project on quarkus 2.6 (planning to upgrade to quarkus 3.x) and we had the same issue : 
`I/O Error : Socket read interrupted` 

@chileme88 @aniash007 @LarsKoelpin @ccarv  Actually in our deployment scenario, the workaround provided by @solerman worked out : 
`-Doracle.jdbc.javaNetNio=false` 
(The text posted by @solerman, as far I can understand,  is an oracle KB that is only accessible with subscription).

Good luck

**Reply by LarsKoelpin** [reply]
The workaround works in general works, but the workaround does not work with the current drivers (see exception above), which where shipped by keycloak/quarkus (oracle version > 21.5.0.0).

However, in my specific case I was able to resolve it by providing an old version (ojdbc 21.5.0.0) of the oracle-driver in the providers/ directory of my keycloak instance (and upgrade from 22.0.04 to 23.0.2).

**Comment by fugerit79** [comment]
Thanks @LarsKoelpin , when we upgrade to quarkus 3 we'll neet to take it into consideration (and it will be soon as [quarkus 2.X is at end of life](https://quarkus.io/security/))

**Comment by luisgambao** [comment]
-Doracle.jdbc.javaNetNio=false did not work for me. However, setting a high quarkus.datasource.jdbc.acquisition-timeout did. Turns out that (at least in my case) the Socket read is interrupted because of the timeout. That's why I think some queries fail, up until a connection is made and then you never have the same issue again.

**Comment by saadbenrouyne** [comment]
Hi everyone,

I just came across this issue.

My config is :
- Keycloak 26.0.9.redhat-00001 on JVM (powered by Quarkus 3.15.1.redhat-00005)
- Oracle 19c 

Using -Doracle.jdbc.javaNetNio=false with the latest ojdb driver [23.7.0.25.01](https://mvnrepository.com/artifact/com.oracle.database.jdbc/ojdbc11/23.7.0.25.01) (as of now) worked for me.

The following drivers did not work me :

- ojdbc11-23.2.0.0 : 19c (19.3.0) (Oracle JDBC Driver v23.2.0.0) This driver is tested and certified by Redhat for Red Hat build of Keycloak 26.0.x but not working in my case. In fact, Keycloak doesn't even start when using -Doracle.jdbc.javaNetNio=false. Similar exceptions to the ones reported  by @LarsKoelpin occur.

- ojdbc11-23.5.0.24.07: This driver is the one supported by Keycloak for Oracle 19c according to Keycloak's official documentation, but it did not work in my case. This driver does not fire any exceptions on Keycloak startup, however it appears to ignore the -Doracle.jdbc.javaNetNio=false property.

I guess some drivers have bugs from what I understand.

Hope it helps. Good luck !

**Reply by mahi-kandiar** [reply]
For us adding ?oracle.jdbc.javaNetNio=false int the jdbc url helped. no jdbc version change
